src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -399,19 +409,25 @@ class FastSync(
       val hashesToRequest = (nodeData.values.indices zip receivedHashes) flatMap { case (idx, valueHash) =>
         requestedHashes.find(_.v == valueHash) map {
           case _: StateMptNodeHash =>
-            handleMptNode(nodeData.getMptNode(idx))
+            if(!(syncState.checkDB && blockchain.getMptNodeByHash(valueHash).isDefined)){
+              handleMptNode(nodeData.getMptNode(idx))
+            } else Nil
 
           case _: ContractStorageMptNodeHash =>
-            handleContractMptNode(nodeData.getMptNode(idx))
+            if(!(syncState.checkDB && blockchain.getMptNodeByHash(valueHash).isDefined)) {
```

Comment:
this if structore could be extracted to separate method, because it is used multiple times